Speed up reductions on non-contiguous dimensions

adamlerer opened the pull request:
I just did `sum`, `prod`, `mean`, `max`, and `min` for now, but you can do a bunch of the other pointwise reductions like this too. 

before:
```
th> t = torch.randn(64,1024,1024)
                                                                      [5.6855s]
th> t:max(1);
                                                                      [1.7969s]
th> t:max(2);
                                                                      [0.9834s]
th> t:max(3);
                                                                      [0.1487s]
th> t:sum(1);
                                                                      [1.5895s]
th> t:sum(2);
                                                                      [0.9665s]
th> t:sum(3);
                                                                      [0.1610s]
```

after:
```
th> t = torch.randn(64,1024,1024)
                                                                      [5.7820s]
th> t:max(1);
                                                                      [0.3097s]
th> t:max(2);
                                                                      [0.2392s]
th> t:max(3);
                                                                      [0.1456s]
th> t:sum(1);
                                                                      [0.1985s]
th> t:sum(2);
                                                                      [0.2219s]
th> t:sum(3);
```

I think the optimizations for `max` and `min` will conflict with the TH_TENSOR_APPLY improvements in https://github.com/torch/torch7/pull/852, but we can always add special macros for this purpose if we want to keep the max and min optimizations. That new macro could also support stuff like `cumsum`.

soumith commented:
it looks like we're lucky that TH_TENSOR_DIM_APPLY2 etc. dont have OpenMP optimizations yet. in the longer term, we need tree reductions.

fmassa commented on lib/TH/generic/THTensorMath.c:
```diff
@@ -1124,32 +1129,70 @@ void THTensor_(max)(THTensor *values_, THLongTensor *indices_, THTensor *t, int
+      return;
+    }
+
+    THTensor *tempValues_ = THTensor_(newWithTensor)(values_);
```
Don't you need to free this tensor? Same for the indices.

adamlerer replied:
Yep, thanks. 👍

apaszke commented on lib/TH/generic/THTensorMath.c:
```diff
@@ -1124,32 +1129,70 @@ void THTensor_(max)(THTensor *values_, THLongTensor *indices_, THTensor *t, int
+  } else {
+    if (THTensor_(nDimension)(t) > 1) {
+      THTensor *t0 = THTensor_(newSelect)(t, dimension, 0);
+      THTensor_(copy)(values_, t0);
```
An optimization would be to fill it with the smallest possible value for each type. You're going to compare the values with the first slice anyway, and that would save some bandwidth.

adamlerer replied:
That seems like a pretty annoying optimization just to save 1/size[dim] bandwidth. If you really wanted to, the clean way would be to initialize the result with t[0] and then do the APPLY over t[1:].

apaszke commented on lib/TH/generic/THTensorMath.c:
```diff
@@ -1124,32 +1129,70 @@ void THTensor_(max)(THTensor *values_, THLongTensor *indices_, THTensor *t, int
+    // tempIndices has stride 0 and size >1 in dimension, there will definitely
+    // be an instantiated counter dimension there; this might not be true after
+    // some new optimizations to TH_TENSOR_APPLY3, so we'll need a different
+    // set of macros.
```
The problem arises from the fact that apply macros will merge the counters, right?

adamlerer replied:
Yes, that's correct. It's unclear what the contract is with TH_TENSOR_APPLY... we may need one set of macros that exposes the counter and one that is allowed to optimize it away.

apaszke commented on lib/TH/generic/THTensorMath.c:
```diff
@@ -1192,12 +1277,24 @@ void THTensor_(sum)(THTensor *r_, THTensor *t, int dimension)
+    temp_->size[dimension] = t->size[dimension];
+    temp_->stride[dimension] = 0;
+
+    THTensor_(cadd)(temp_, temp_, 1, t);
```
Are you sure that is correct? I'd be very careful with mutating operations on expanded tensors. Parallelizing `cadd` will lead to race conditions here.

adamlerer replied:
Good question. I know it's currently correct, but the question is whether it relies on the implementation. It comes down to whether `cadd` behavior is defined for destinations with stride-0 dimensions. Currently it does (at least for the CPU), because it only parallelizes the contiguous case.

Perhaps it would be cleaner to use `TH_APPLY` here directly instead of `cadd`, because we can require that `TH_APPLY` handles stride-0 destination properly. If we want to parallelize it in the future we'll need to split it into two versions.

apaszke replied:
Even if it's correct at the moment, I'm afraid that we'll change `cadd` in the future, and have other seemingly unrelated things break. Not sure about the remaining part, I have to look into the apply macros to see what they do in detail.

adamlerer replied:
Sure, I'll change it to use TH_APPLY directly.

If someone changes `TH_APPLY` to no longer handle stride-0 dimensions properly, the tests will fail and they'll have to fix it :P

apaszke replied:
If it's going to be a loud fail then that sounds good to me, but I'd rather save time on debugging race conditions in the future 😄

adamlerer commented:
@soumith re OpenMP: as I said to @apaszke this comes down to the specification of `TH_TENSOR_APPLY` (not TH_TENSOR_DIM_APPLY) w.r.t. stride-0 dimensions. Currently it handles them correctly. If someone changes it in a way that breaks that, they'll need to maintain a version that handles stride-0 correctly. Note that there's already existing code that sort of depends on handling this case correctly, e.g.

```
t = torch.Tensor(10, 10, 10)
u = torch.Tensor(1, 10, 10).expand_as(t).zero_()
u.add_(t)
```

so I'm not sure if it is now safe to break that (considering the specification was ill-defined).

fmassa commented on lib/TH/generic/THTensorMath.c:
```diff
@@ -2398,35 +2401,11 @@ void THTensor_(lerp)(THTensor *r_, THTensor *a, THTensor *b, real weight)
-    THTensor_(free)(temp_);
-    THTensor_(div)(r_, r_, t->size[dimension]);
-  }
+  THTensor_(sum)(r_, t, dimension);
```
Just wondering, is there a performance penalty in the contiguous case with this change?

apaszke commented on lib/TH/generic/THTensorMath.c:
```diff
@@ -1124,32 +1129,73 @@ void THTensor_(max)(THTensor *values_, THLongTensor *indices_, THTensor *t, int
+    } else {
+      THTensor_(fill)(values_, THTensor_(get1d)(t, 0));
+    }
+    THLongTensor_zero(indices_);
```
You need to initialize them with `TH_INDEX_BASE` (applies to other functions too).

apaszke replied:
Also, it would be good to extend the tests to cover this.

adamlerer replied:
This is not necessary, because cwrap adds 1 to the indexes in Lua. I'll add the test though.

apaszke replied:
Right. I forgot about that.

Jokeren commented on lib/TH/generic/THTensorMath.c:
```diff
@@ -1124,32 +1129,73 @@ void THTensor_(max)(THTensor *values_, THLongTensor *indices_, THTensor *t, int
+    // be an instantiated counter dimension there; this might not be true after
+    // some new optimizations to TH_TENSOR_APPLY3, so we'll need a different
+    // set of macros.
+    TH_TENSOR_APPLY3(real, t, real, tempValues_, long, tempIndices_,
```
Hi, @adamlerer , could you explain why `t` and `tempValues_` has the same size? I notice you have comment `tempIndices_.expand_as(t)`.

Jokeren commented on lib/TH/generic/THTensorMath.c:
```diff
@@ -1124,32 +1129,73 @@ void THTensor_(max)(THTensor *values_, THLongTensor *indices_, THTensor *t, int
+    TH_TENSOR_APPLY3(real, t, real, tempValues_, long, tempIndices_,
+                     if(!(*t_data <= *tempValues__data) && !th_isnan(*tempValues__data)) {
+                       *tempValues__data = *t_data;
+                       *tempIndices__data = tempIndices__counter[dimension];
```
The problem is that `dimension` works only in the raw counter, not the compressed one. So I am afraid that you must use another macro for it, or extra dimension calculations are needed.

adamlerer replied:
Yes I know. So what should we do here? I can either remove the `min` and `max` optimizations and save them for a later pass with new macros, or I can copy-paste the old `TH_TENSOR_APPLYX` macros just for min and max in this PR. Thoughts? @apaszke

Jokeren replied:
I would vote for using the old `TH_TENSOR_APPLYX`.

apaszke replied:
I think we could leave them as TH_TENSOR_APPLY_COUNTER or sth like that. They're all unacceptably slow, so let's keep the fix.

adamlerer replied:
Will do.

apaszke commented:
I think we could leave them as TH_TENSOR_APPLY_COUNTER or sth like that. They're all unacceptably slow, so let's keep the fix.

adamlerer commented:
Alright, I rebased this on top of https://github.com/torch/torch7/pull/852, refactored TH_TENSOR_APPLY to remove all the copypasta and added a variant that can handle these reduction cases.

You'll have to wait until https://github.com/torch/torch7/pull/852 is landed to review this.

adamlerer commented on lib/TH/THTensorApply.h:
```diff
@@ -595,8 +131,90 @@
 
+#define TH_TENSOR_APPLY(TYPE, TENSOR, CODE) \
+  TH_TENSOR_APPLY_D(TYPE, TENSOR, -1, CODE)
+
```
To make reviewing easier, note that most of the changes to this file are just a refactoring to remove copy-pasted macros. The only functional changes, which was necessary for the reduction operations, can be found by searching for uses of the new `DIM` argument to `TH_TENSOR_APPLYX_D`

soumith commented:
thanks Adam!
